File: C++/90 Subsets II.cpp
class Solution
{
public:
    vector<vector<int>> subsetsWithDup(vector<int> &nums)
    {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        vector<int> path;
        auto dfs = [&](this auto &&dfs, int u) -> void
        {
            if (u == n)
            {
                ans.push_back(path);
                return;
            }

            int x = nums[u];
            path.push_back(x);
            dfs(u + 1);
            path.pop_back();

            int k = u + 1;
            while (k < n && nums[k] == x)
            {
                k++;
            }

            dfs(k);
        };

        dfs(0);
        return ans;
    }
};

File: Swift/90. Subsets II.swift
class Solution {
    func subsetsWithDup(_ nums: [Int]) -> [[Int]] {
        let res = nums.sorted()
        var ans: [[Int]] = []
        var path: [Int] = []
        let n = nums.count
        func dfs(_ u: Int) {
            if u == n {
                ans.append(path)
                return
            }

            let x = res[u]
            path.append(x)
            dfs(u + 1)
            path.popLast()

            var k = u + 1
            while k < n, res[k] == x {
                k += 1
            }
            dfs(k)
        }
        dfs(0)
        return ans
    }
}
